source/Variant.cpp

```diff
@@ -0,0 +1,306 @@
+		return true;
+
+	return false;
+}
+
+
+
+// Choose a list of ships from this variant. All ships from the ships
+// vector are chosen, as well as a random selection of ships from any
+// nested variants in the WeightedList.
+vector<const Ship *> Variant::ChooseShips() const
+{
+	vector<const Ship *> chosenShips = ships;
+	for(const auto &it : variants)
+		for(int i = 0; i < it.Weight(); ++i)
+			chosenShips.push_back(it.Get().NestedChooseShip());
+	return chosenShips;
+}
+
+
+
+// The strength of a variant is the sum of the cost of its ships and
+// the strength of any nested variants.
+int64_t Variant::Strength() const
+{
+	int64_t sum = 0;
+	for(const Ship *ship : ships)
+		sum += ship->Cost();
+	for(const auto &variant : variants)
+		sum += variant.Get().NestedStrength() * variant.Weight();
+	return sum;
+}
+
+
+
+// An operator for checking the equality of two variants.
+bool Variant::operator==(const Variant &other) const
+{
+	// Is either variant named? Do they share the same name?
+	if(!other.name.empty() || !name.empty())
+		return other.name == name;
+
+	// Are the ships of other a permutation of this variant's?
+	if(other.ships.size() != ships.size()
+		|| !is_permutation(ships.begin(), ships.end(), other.ships.begin()))
+		return false;
+	// Are the nested variants of other a permutation of this variant's?
+	if(other.variants.size() != variants.size()
+		|| !is_permutation(variants.begin(), variants.end(), other.variants.begin()))
+		return false;
+
+	// If all checks have passed, these variants are equal.
+	return true;
+}
+
+
+
+// An operator for checking the inequality of two variants.
+bool Variant::operator!=(const Variant &other) const
+{
+	return !(*this == other);
+}
+
+
+
+// Check whether a variant is contained within itself.
+bool Variant::NestedInSelf(const string &check) const
+{
+	if(!name.empty() && name == check)
+		return true;
+
+	for(const auto &it : variants)
+		if(it.Get().NestedInSelf(check))
+			return true;
+
+	return false;
+}
+
+
+
+// Determine if this variant template uses well-defined data as a
+// nested variant.
+bool Variant::NestedIsValid() const
+{
+	// All possible ships must be valid.
+	if(any_of(ships.begin(), ships.end(),
+			[](const Ship *const s) noexcept -> bool { return !s->IsValid(); }))
+		return false;
+
+	// All possible nested variants must be valid.
+	if(any_of(variants.begin(), variants.end(),
+			[](const WeightedVariant &v) noexcept -> bool { return !v.Get().NestedIsValid(); }))
+		return false;
+
+	return true;
+}
+
+
+
+// Choose a ship from this variant given that it is a nested variant.
+// Nested variants only choose a single ship from among their list
+// of ships and variants.
+const Ship *Variant::NestedChooseShip() const
+{
+	int total = ships.size() + variants.TotalWeight();
+	// Randomly choose between the ships and the variants.
+	if(static_cast<int>(Random::Int(total)) < static_cast<int>(variants.TotalWeight()))
```

Comment:
Random::Int returns a uint32_t and WeightedVariant::TotalWeight returns a size_t. If I recall correctly my compiler threw an error on comparing those two, hence why I casted them both to ints.

Reply:
I spotted that the only other place a Random::Int call is directly compared against anything that isn't another Random::Int call, at least so far as I could find, was a comparison between Random::Int and explosionCount in Ship.cpp. Since explosionCount is an `unsigned`, I've changed this to store variants.TotalWeight() in an `unsigned` variable.